tools/oeedger8r/Emitter.ml

```diff
@@ -451,30 +445,45 @@ let warn_size_and_count_params (fd : func_decl) =
 
 (** Generate the Enclave code. *)
 let gen_enclave_code (ec : enclave_content) (ep : edger8r_params) =
-  (* We need to check Ptrs for Foreign or Struct types, then check
-    those against the user's structs, and then check if any members
-    should be deep copied. *)
-  let get_deepcopy a structs ep =
-    let get_struct = function
-      | Ptr a -> (
-        match a with
-        | Foreign n | Struct n ->
-            (* TODO: This would simply be [List.assoc_opt] if we had at
-             least 4.05. *)
-            if List.mem_assoc n structs then Some (List.assoc n structs)
-            else None
-        | _ -> None )
+  (* We need to check [Ptr]s for [Foreign] or [Struct] types, then
+     check those against the user's [Struct]s, and then check if any
+     members should be deep copied. What we return is the list of
+     members of the [Struct] which should be deep-copied, otherwise we
+     return an empty list. *)
+  let get_deepcopy a =
+    (* Check if [a] is a [Ptr], then check if the [Ptr] is a
+       user-defined [Foreign] or [Struct], then return the associated
+       [StructDef], else return [None]. *)
+    let should_deepcopy_a = function
+      | Ptr p ->
+          (* [ec.comp_defs] is a list of all composite types, but we're
+             only interested in the structs, so we filter out the rest
+             and unwrap them from [composite_type]. *)
+          let structs =
+            filter_map
+              (function StructDef s -> Some s | _ -> None)
+              ec.comp_defs
+          in
+          let is_match name s = s.sname = name in
```

Comment:
@anakrish Can this be cleaned up? Without `List.assoc_opt` or `List.find_opt`, this is about the best I could do.

Reply:
Reading the code, the 'structs' expression depends only on ec.comp_defs, so it could be moved outside of get_deepcopy and thus be computed only once.
There is no guarantee that Ocaml itself will move the evaluation of 'structs' outside get_deepcopy.

I feel moving it out would make it more readable.
Also have a function get_struct_by_name that looks up the structs given a name.

Reply:
With a get_struct_by_name function, it seems like should_deepcopy_a can be made more declarative:

let shoud_deepcopy_a = function
| Ptr (Struct n) -> get_struct_by_name n
| Ptr (Foreign n) -> get_struct_by_name n
| -> None

I would suggest renaming this function since the function name suggests a bool return type contrary to the implementation

Reply:
Omg I forgot about nesting pattern matching. TYVM.

Reply:
> Reading the code, the 'structs' expression depends only on ec.comp_defs, so it could be moved outside of get_deepcopy and thus be computed only once.
> There is no guarantee that Ocaml itself will move the evaluation of 'structs' outside get_deepcopy.
> 
> I feel moving it out would make it more readable.

I feel like this is just premature optimization though. We're certainly not running up against noticeable runtimes with the edger8r. I'll make some changes here, but going for readability over performance optimizations.

Reply:
I'm more interested in readability and scoping functions appropriately would make it more readable.